### PR TITLE
Add OpenStack Zun provider(#22)

### PR DESCRIPTION
**Pull request by kevin-zhaoshuai**
Add OpenStack Zun provider.

Zun primarily uses the concepts of Container and Capsule. A container is a single container backed by Docker or other container engines. The basic Docker primitives, such as create/delete container, are integrated with Zun. In addition, Zun adds several OpenStack-integrated capabilities to containers. For example, a Zun container is, by default, given Neutron IP addresses, has an option to attach Cinder volumes, the access is authenticated by Keystone and several other services. OpenStack users already familiar with the OpenStack concepts will find it fairly easy to get started with Zun containers.

In addition to the Container concept, Zun provides some slightly advanced abstractions such as Capsule. A Zun Capsule, like a Kubernetes Pod, represents a group of containers that are co-located, co-scheduled and have some shared contexts. Capsule is used to group multiple containers that need to work closely with each other to achieve a business goal.

The operational requirements of Zun are similar to other OpenStack services. They require installing a Zun API server at the controller node and a Zun compute agent at each compute node. The primary dependency of Zun is the set of OpenStack Base Services: an Oslo compatible database, message queue, Etcd and Keystone.

### CODE REVIEW

**Comment by hongbin** [comment]
@dims FYI

**Comment by kevin-zhaoshuai** [comment]
recheck

**Comment by hongbin** [comment]
Here is the error:

providers/openstack/zun.go:310:60: capsule.Containers[1].StartedAt undefined (type capsules.Container has no field or method StartedAt)
providers/openstack/zun.go:424:43: cs.StartedAt undefined (type *capsules.Container has no field or method StartedAt)
providers/openstack/zun.go:436:44: cs.StartedAt undefined (type *capsules.Container has no field or method StartedAt)

It complained the missing of 'StartedAt' field, but it looks this field has been added at https://github.com/gophercloud/gophercloud/commit/57beb4b85593615e5a3998e89f4b87a8351fb428#diff-fccc6ed9a59a35d5886f7468f7842ca5 . Any idea?

**Comment by hongbin** [comment]
@dims are you familiar with the dependency management in golang. If yes, would appreciate your advice on the CI issue of this PR. Thanks in advance.

**Comment by hongbin** [comment]
recheck

**Comment by hongbin** [comment]
I looks the CI were not successful because it ran at 17 ago. At that time, the gophercloud PR [1] was not merged yet.

[1] https://github.com/gophercloud/gophercloud/pull/1055

**Comment by kevin-zhaoshuai** [comment]
recheck

**Comment by dims** [comment]
@kevin-zhaoshuai @hongbin please see 3 more commits that are needed to get the dependencies right and get the CI to go green here - https://github.com/virtual-kubelet/virtual-kubelet/pull/257/commits

**Comment by kevin-zhaoshuai** [comment]
@dims Thanks a lot! I've submit the patch to this PR to see the CI results.

**Comment by kevin-zhaoshuai** [comment]
@rbitia @dims @robbiezhang  sorry for late respond. I just finished fix all the conflict.

**Comment by cpuguy83** [comment]
Is this still WIP? Looks like you need to regenerate Gopkg.lock as well.

**Comment by kevin-zhaoshuai** [comment]
@rbitia @dims @robbiezhang ping for review

**Comment by dims** [comment]
@kevin-zhaoshuai looks good! we should setup some automated testing using https://openlabtesting.org too

@mrhillsman @hongbin  can you please help with adding some CI around this?

**Comment by cpuguy83** [comment]
You'll need to register this provider in `providers/register`. Please make sure to provide a build tag that allows people to omit the provider (you can follow the examples in all the other providers there).

**Comment by mrhillsman** [comment]
@dims ack - would def re @hongbin assistance i believe - will review tomorrow

File: providers/register/provider_openstack.go [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package register
```
Comment (cpuguy83) [inline providers/register/provider_openstack.go]: Can you add a build tag to this?

```suggestion
// +build !no_openstack_provider

package register
```

**Comment by kevin-zhaoshuai** [comment]
@cpuguy83 ping   for review, thanks

**Comment by kevin-zhaoshuai** [comment]
@mrhillsman   ping  for review. Thanks

File: providers/openstack/zun.go [hunk 2]
```diff
@@ -0,0 +1,509 @@
...
+
+	podCreationTimestamp = metav1.NewTime(capsule.CreatedAt)
+	if len(capsule.Containers) > 1 {
+		containerStartTime = metav1.NewTime(capsule.Containers[1].StartedAt)
+	}
+	containerStartTime = metav1.NewTime(time.Time{})
+	// Deal with container inside capsule
+	containers := make([]v1.Container, 0, len(capsule.Containers))
+	containerStatuses := make([]v1.ContainerStatus, 0, len(capsule.Containers))
+	// First container is sandbox
```
Comment (hongbin) [inline providers/openstack/zun.go]: This is not true anymore. The sandbox container is removed from the list. All the containers in the list are the actual containers declared in the pod.

File: providers/openstack/zun.go [hunk 3]
```diff
@@ -0,0 +1,509 @@
...
+	for _, c := range capsule.Containers[1:] {
```
Comment (hongbin) [inline providers/openstack/zun.go]: As commented above, no need to skip the first container anymore

File: providers/openstack/zun.go [hunk 4]
```diff
@@ -0,0 +1,509 @@
...
+	return providers.OperatingSystemLinux
+}
+
+func capsuleToPod(capsule *capsules.Capsule) (*v1.Pod, error) {
+	var podCreationTimestamp metav1.Time
+	var containerStartTime metav1.Time
+
+	podCreationTimestamp = metav1.NewTime(capsule.CreatedAt)
+	if len(capsule.Containers) > 1 {
+		containerStartTime = metav1.NewTime(capsule.Containers[1].StartedAt)
```
Comment (hongbin) [inline providers/openstack/zun.go]: I guess we can change to:

```
	if len(capsule.Containers) > 0 {
		containerStartTime = metav1.NewTime(capsule.Containers[0].StartedAt)
```

File: providers/openstack/zun.go [hunk 5]
```diff
@@ -0,0 +1,509 @@
...
+	case "Deleting":
+		return v1.PodPending
+	case "Deleted":
+		return v1.PodPending
+	}
+
+	return v1.PodUnknown
+}
+
+func zunCapStatusToPodPhase(status string) v1.PodPhase {
```
Comment (hongbin) [inline providers/openstack/zun.go]: In overall, with the latest API version, container and capsule status would be the same, so I guess we can remove this function and re-use zunStatusToPodPhase for converting the capsule's status to pod.

**Comment by hongbin** [comment]
@mrhillsman @dims ACK. I will add the CI for this provider.

**Comment by hongbin** [comment]
All, I think this patch is ready for review. I will work on setting up the CI, which could be in another PR.

File: providers/register/provider_openstack.go [hunk 6]
```diff
@@ -0,0 +1,21 @@
+// +build !no_openstack_provider
```
Comment (cpuguy83) [inline providers/register/provider_openstack.go]: Sorry, we ended up flipping this around. Can you change the build tag to `openstack_provider`

Comment (hongbin) [reply providers/register/provider_openstack.go]: Done

**Comment by hongbin** [comment]
> 1 issue (sorry for changing this!) but otherwise LGTM.
> 
> Please note that we do plan on changing some of these provider interfaces (#520) and split out providers out of tree (#494)

ACK. Thanks for the information. Both proposals sound OK for me.

**Pull request merged**